stdlib/builtins.pyi

```diff
@@ -940,24 +939,43 @@ class bool(int):
     def __invert__(self) -> int: ...
 
 @final
-class slice(Generic[_StartT, _StopT, _StepT]):
+class slice(Generic[_StartT_co, _StopT_co, _StepT_co]):
     @property
-    def start(self) -> _StartT: ...
+    def start(self) -> _StartT_co: ...
     @property
-    def step(self) -> _StepT: ...
+    def step(self) -> _StepT_co: ...
     @property
-    def stop(self) -> _StopT: ...
-    @overload
-    def __new__(cls, stop: int | None, /) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
-    @overload
-    def __new__(
-        cls, start: int | None, stop: int | None, step: int | None = None, /
-    ) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
-    @overload
-    def __new__(cls, stop: _T2, /) -> slice[Any, _T2, Any]: ...
-    @overload
-    def __new__(cls, start: _T1, stop: _T2, /) -> slice[_T1, _T2, Any]: ...
-    @overload
+    def stop(self) -> _StopT_co: ...
+    # unary overloads ------------------------------------------------------------------
+    @overload  # 1x None
+    def __new__(cls, stop: None, /) -> slice: ...
+    @overload  # 0x None
+    def __new__(cls, stop: _T2, /) -> slice[_T2 | None]: ...
+    # binary overloads -----------------------------------------------------------------
+    @overload  # 2x None
+    def __new__(cls, start: None, stop: None, /) -> slice: ...
+    @overload  # 1x None
+    def __new__(cls, start: _T1, stop: None, /) -> slice[_T1 | None]: ...
+    @overload  # 1x None
+    def __new__(cls, start: None, stop: _T2, /) -> slice[_T2 | None]: ...
+    @overload  # 0x None
+    def __new__(cls, start: _T1, stop: _T2, /) -> slice[_T1, _T2]: ...
+    # tenary overloads -----------------------------------------------------------------
+    @overload  # 3x None
+    def __new__(cls, start: None, stop: None, step: None, /) -> slice: ...
+    @overload  # 2x None
+    def __new__(cls, start: None, stop: None, step: _T3, /) -> slice[Any, Any, _T3]: ...
+    @overload  # 2x None
+    def __new__(cls, start: _T1, stop: None, step: None, /) -> slice[_T1 | None, _T1 | None]: ...
+    @overload  # 2x None
+    def __new__(cls, start: None, stop: _T2, step: None, /) -> slice[_T2 | None, _T2 | None]: ...
+    @overload  # 1x None
+    def __new__(cls, start: _T1, stop: None, step: _T3, /) -> slice[_T1 | None, _T1 | None, _T3]: ...
+    @overload  # 1x None
+    def __new__(cls, start: None, stop: _T2, step: _T3, /) -> slice[_T2 | None, _T2 | None, _T3]: ...
+    @overload  # 1x None
+    def __new__(cls, start: _T1, stop: _T2, step: None, /) -> slice[_T1, _T2]: ...
```

Comment:
This could be combined with your fourth binary overload:

```py
@overload
def __new__(cls, start: _T1, stop: _T2, step: None = None, /) -> slice[_T1, _T2]: ...
```